### ISSUE TITLE
docker --help fails if your key.json has a problem

### ISSUE THREAD

**Issue by SvenDowideit** [docker/docker#10297]
I've been trying to run more than one docker daemon on the one host, and this finally leads to a bad config, which then means I can't even get cmdline help

this is the 1.5.0rc1

```
[sven@t440s docker-swarm]$ docker --help
FATA[0000] Error loading key file: unable to decode private key JWK: JWK EC Public Key ID does not match: 46KY:OYL4:YZUU:HSN7:SSQH:ZPVI:63KX:PCYE:LN2R:6XO3:WWBE:HW7E 
[sven@t440s docker-swarm]$ docker
FATA[0000] Error loading key file: unable to decode private key JWK: JWK EC Public Key ID does not match: 46KY:OYL4:YZUU:HSN7:SSQH:ZPVI:63KX:PCYE:LN2R:6XO3:WWBE:HW7E 
[sven@t440s docker-swarm]$ docker --version
Docker version 1.5.0, build 7e803ba
```

**Comment by SvenDowideit** [comment]
deleteing the `~/.docker/key.json` fixes it - but 

- the error doesn't help the user know that
- the error should not prevent the help text.

**Comment by jfrazelle** [comment]
ping @dmcgowan

**Comment by dmcgowan** [comment]
This key was originally generated by a dev version before the 1.4 release.  The fix is as you mentioned, delete the key or the KID field from the key. It is was a known issue when we changed the algorithm but it happened on master before an official release.  If this error re-occurs can you please list the versions of the docker daemons you are running.

**Comment by duglin** [comment]
Aside from dealing with a bad json file, I think part of the issue that needs to be fixed is that the help text being displayed is dependent on reading the json file - that's a bit odd.  No matter how bad things are, you'd think "docker --help" should pretty much always work.

From looking at the code it appears that we have quite a bit of security code running prior to getting to the CmdHelp() function. I wonder if it would be possible to move the security code to someplace later - or perhaps treat the non-error "help" case as special and check for that prior to the security code.

**Comment by swernli** [comment]
It seems that this file is also preventing docker client from starting on Windows.  I get:
←[31mFATA←[0m[0000] Error saving key file: unable to write private key file C:\Users\swernli\.docker\key.json: open C:\Users\swernli\.docker\key.json: The system cannot find the path specified.

Stepping through the code, it appears that call os.OpenFile with the flags "os.O_WRONLY|os.O_CREATE|os.O_TRUNC" does not correctly create the file on Windows if it doesn't already exist.  Is this a known bug?

**Comment by tianon** [comment]
On 23 January 2015 at 08:17, Derek McGowan <notifications@github.com> wrote:

> It is was a known issue when we changed the algorithm but it happened on
> master before an official release.


Can we please not do this kind of shortcut in the future?  Historically,
"master" as always been releasable, and it's a really great tradition since
it encourages people to actually use and test it.

**Comment by jfrazelle** [comment]
+1 to what @tianon, master is always supposed to be stable, there will **not** be any testing on master.

**Comment by swernli** [comment]
Looks like the issue on Windows is actually a problem of incorrect use of "path" instead of "filepath" when dealing with filesystem paths.  I'll create a separate issue for this.

**Comment by icecrime** [comment]
Reading through the code, allowing `--help` in that case would probably be an ugly hack. My opinion here is that #10360 is good enough: wdyt? (ping @crosbymichael)

**Comment by crosbymichael** [comment]
Yes, i think #10360 is a valid fix for this.  Its seems like a very isolated issue right now.  The path/filepath problems where fixed by #10323 reported by another user.  

Everyone needs to remember that master is always releasable so never merge anything that you are not comfortable with going out in a release but I think we can say that this issue is resolved.
